[wenenhoe] iP

wenenhoe opened the pull request.

wenenhoe committed:
style: Rename chatbot

Give chatbot a new name to differentiate it from the placeholder name
`Duke`

wenenhoe committed:
feat: Add greetings and exit

Implement initial skeletal version that greets user and exits

wenenhoe committed:
feat: Add echo functionality

Echo command entered by user and exit when user types `bye`

wenenhoe committed:
feat: Add task storing and listing

Store whatever text entered by user and display back when requested

wenenhoe committed:
style: Improve readability and interactivity of chatbot

New stylings added:
* Separator
* Tab
* Console-like prompt

wenenhoe committed:
bugfix: Invert filter for user input

Filter `list` and `bye` first then allow everything else to be added to
task list

wenenhoe committed:
refactor: Abstract task into separate class

wenenhoe committed:
refactor: Utilise switch case to filter user input

wenenhoe committed:
feat: Add mark and unmark functionality

Allow user to:
* Mark tasks as done
* Change tasks back to not done

wenenhoe committed:
refactor: Abstract task management

* Print task list
* Add task
* Mark task as done
* Unmark task as not done
* Parse user input for task number

wenenhoe committed:
feat: Add exception for parsing user input

* Handle invalid task number
* Handle out of range task number

wenenhoe committed:
style: Modify code to follow coding conventions

* Inline method annotation
* Descriptive variable naming
* Comment for ignored exception

wenenhoe committed:
fix: Add static modifier to method

Method independent of internal state of class

wenenhoe committed:
fix: Remove unnecessary type cast

Object of interface type `List<>` can be iterated without type cast to
`Array`

wenenhoe committed:
style: Add newline for `Override` annotation

wenenhoe committed:
feat: Add new subclasses for different Task types

* Deadline
* Event
* Todo

wenenhoe committed:
refactor: Abstract console input and output

* Print welcome message
* Print formatted response
* Get user input from console with prompt

wenenhoe committed:
refactor: Modify `TaskManager` methods

* Methods take in task number instead of string that needs to be parsed
* Move parsing methods into `Magus` class

wenenhoe committed:
fix: Apply try catch at relevant code statements

Only used when trying to get a `Task` at index, instead of irrelevant
code statements

wenenhoe committed:
refactor: Reorganise code into packages

* task related classes
* console related class

wenenhoe committed:
feat: Add `Command` enum

* Allow improved switch case processing of commands
* Improve scalability for future support of new commands

wenenhoe committed:
refactor: Utilise `Command` enum in switch case

Improve processing of commands from user input

wenenhoe committed:
refactor: Abstract get task into method

Remove duplicate code in `markTaskAsDone` and `unmarkTaskAsNotDone`

wenenhoe committed:
feat: Handle new commands for different task types

* todo
* deadline
* event

wenenhoe committed:
refactor: Utilise `Console.printResponse`

For printing to console with tabs

wenenhoe committed:
feat: Add parsers for each task type

Handle additional arguments for each task type:
* Deadline `by`
* Event `from` and `to`

wenenhoe committed:
fix: Remove unnecessary `toString` method calls

wenenhoe committed:
refactor: Improve processing commands readability

Break into smaller methods

wenenhoe committed:
refactor: Abstract user input parsing

Parse command and additional input

wenenhoe committed:
fix: Resolve `NoSuchElementException`

* Encounter issue when `Scanner` is recreated
* Should be static and created once for `System.in` stream

wenenhoe committed:
test: Modify runtest scripts to run `Magus`

wenenhoe committed:
test: Add test cases and expected output

syj02 commented on src/main/java/console/Parser.java:
```diff
@@ -0,0 +1,59 @@
+package console;
```
I like the packaging of your classes. Makes it a lot neater and easier to read

syj02 commented on src/main/java/task/TaskManager.java:
```diff
@@ -0,0 +1,88 @@
+        try {
+            return taskList.get(taskNum);
+        } catch (IndexOutOfBoundsException ok) {
+            return null; // Outside range of task list, unable to modify task status
```
Maybe can remove "unable to modify task"? "Outside range of task list" is sufficient to understand.

wenenhoe replied:
Thank you for your comment. This has been resolved in 17ee986.

syj02 commented on src/main/java/Magus.java:
```diff
@@ -0,0 +1,68 @@
+        }
+    }
+
+    public static boolean processInput(Parser parser, TaskManager taskManager) {
```
Should this method be less than 30 Lines of Code?

syj02 commented on src/main/java/task/Deadline.java:
```diff
@@ -0,0 +1,36 @@
+
+        List<String> infoList = List.of(taskInfo.split(" "));
+        int commandIndex = infoList.indexOf(commandArg);
+        if (commandIndex == -1) {
```
What does it mean by "commandIndex == -1"? Maybe can include a comment.

wenenhoe replied:
Thank you for your comment. This has been resolved in 17ee986.

Cryolian commented on src/main/java/Magus.java:
```diff
@@ -0,0 +1,68 @@
+            Parser parser = new Parser(input);
+            boolean isContinue = Magus.processInput(parser, taskManager);
+
+            if (!isContinue) {
```
Maybe use isContinue in the while statement, rather than adding a break statement.

wenenhoe replied:
Thank you for your comment. This has been resolved in 0a505b4.

Cryolian commented on src/main/java/console/Console.java:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    public static void printResponse(String response) {
+        printResponse(response, 1);
```
Perhaps use a constant variable instead of 1.

wenenhoe replied:
Thank you for your comment. This issue has been resolved in d5b9beb.

Cryolian commented on src/main/java/console/Parser.java:
```diff
@@ -0,0 +1,59 @@
+        return additionalInput;
+    }
+
+    public boolean isSingleWord() {
```
Maybe change the method name to have a verb at the start, such as checkIfSingleWord.

Cryolian commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,32 @@
+        isDone = true;
+    }
+
+    public void unmarkAsNotDone() {
```
Maybe change the method name to be markAsNotDone, for better clarity.

wenenhoe replied:
Thank you for your comment. This issue has been resolved in d5b9beb.

ChinYanXu commented on src/main/java/Magus.java:
```diff
@@ -0,0 +1,68 @@
+            }
+        }
+    }
+
```
Please change the name of this method to sound more like a boolean.

wenenhoe replied:
Thank you for your comment. This has been resolved in 0a505b4.

ChinYanXu commented on src/main/java/console/Console.java:
```diff
@@ -0,0 +1,35 @@
+        System.out.print("> ");
+        return scanner.nextLine().strip();
+    }
+
```
names of contants such as logo and scanner should be in all caps as they are constants.

wenenhoe replied:
Thank you for your comment. This issue has been resolved in d5b9beb.

ChinYanXu commented on src/main/java/Magus.java:
```diff
@@ -0,0 +1,68 @@
+import console.Console;
```
good and consistent use of indentation.

ChinYanXu commented on src/main/java/Magus.java:
```diff
@@ -0,0 +1,68 @@
+import console.Command;
+import console.Parser;
+import task.TaskManager;
+import task.TaskType;
```
naming of variables and methods are consistent with the coding standard

ChinYanXu commented on src/main/java/Magus.java:
```diff
@@ -0,0 +1,68 @@
+        case DEFAULT:
+            taskManager.addTask(input);
+            break;
+        case LIST:
```
if-else form is consistent with java coding standard

wenenhoe committed:
fix(style): Rename methods and members

* Constant members should be all caps
* Rename method for better clarity
* Naming local variable to remove magic number

wenenhoe committed:
fix(style): Add and fix comments

wenenhoe committed:
refactor: Remove boolean return from method

Remove the need for the method to be renamed by moving boolean check
outside of method

wenenhoe committed:
feat: Deprecate adding of normal task

* Initial: Anything that does not match the command words were added as
normal task to the task list.
* Current: This feature has since been deprecated

wenenhoe committed:
fix(bug): Exit program with invalid command

Additional inputs after using `bye` command should not exit program

wenenhoe committed:
feat: Deprecate using of `Parser.input`

* Initial: Input will be used to add normal task
* Current: No normal task will be created

wenenhoe committed:
wip: Add todo for future error handling

wenenhoe committed:
fix(style): Modify comment

To reflect correct description

wenenhoe committed:
refactor: Rename `Command` enum

From `DEFAULT` to `UNKNOWN`

wenenhoe committed:
refactor: Remove `TaskType.DEFAULT`

wenenhoe committed:
refactor: Remove unused `Parser.input`

wenenhoe committed:
refactor: Remove single word check in `Magus`

* Check performed in `Parser` constructor instead
* If list or bye commands have additional input, then command type will
be set to `Command.UNKNOWN`

wenenhoe committed:
refactor: Remove deprecated normal task creation

Prevent creation of normal task outside of class also

wenenhoe committed:
refactor: Modify return type of subclass `parse`

Utilise implicit upcasting instead

wenenhoe committed:
feat: Add `ArgumentNotFoundException`

Catch when insufficient arguments supplied

wenenhoe committed:
feat: Add `CommandNotFoundException`

Catch when unknown command supplied

wenenhoe committed:
refactor: Use default argument

Optimise code by having a default argument

wenenhoe committed:
style: Improve coding style

Left align colon within `for`

wenenhoe committed:
Merge branch 'branch-Level-5'

wenenhoe committed:
feat: Introduce base package `Magus`

All subpackages are now under the same base package

wenenhoe committed:
Merge branch 'branch-A-Packages'

wenenhoe committed:
feat: Add new command to `delete` tasks

wenenhoe committed:
feat: Add abstract method to get badge

wenenhoe committed:
feat: Implement `Task.getBadge` in `Task.toString`

Remove hardcoded badge and use method to return it instead

wenenhoe committed:
feat: Add `Task.toStoredString` for save to file

Different format string from `Task.toString` for ease of parsing later

wenenhoe committed:
refactor: Move `Magus.processInput` to `CommandManager`

wenenhoe committed:
feat: Add new class to handle program data save

wenenhoe committed:
feat: Add export task list

Save task info into `Magus.txt`

wenenhoe committed:
chore: Exclude `Magus.txt` from `.gitignore`

wenenhoe committed:
Merge branch 'branch-Level-6'

wenenhoe committed:
chore: Move `fileio` package within `task`

Better representation of the usage

wenenhoe committed:
feat: Add get `TaskType` by task badge method

Will be used when parsing stored task info

wenenhoe committed:
feat: Add parser for handling stored data

wenenhoe committed:
style: Add comment to explain empty return

wenenhoe committed:
fix(bug): Add missing column to parse

Parse the `isDone` column

wenenhoe committed:
feat: Add import task list from saved data

* Additional parsers for each task types
* `TaskManager` attempts to import task list at initialization

wenenhoe committed:
fix(bug): Special character used in regex pattern

Escape the delimiter pattern

wenenhoe committed:
refactor: Standardise use of delimiter constant

wenenhoe committed:
Merge branch 'branch-Level-7'

# Conflicts:
#	src/main/java/Magus/Magus.java

wenenhoe committed:
fix(bug): Add export task list after task deletion

wenenhoe committed:
chore: Reorganise `Magus.task` package

Create subpackage `Magus.task.variant` for the 3 task types

wenenhoe committed:
fix: Modify root package name

Conform to lowercase naming convention

wenenhoe committed:
fix(test): Compile and run program broken

Repackaged program require:
* correct folder specified for compilation
* using fully qualified name for the class when running program

Referenced:
* https://stackoverflow.com/a/48882180
* https://stackoverflow.com/a/18093929

wenenhoe committed:
chore: Exclude test `Magus.txt` from `.gitignore`

wenenhoe committed:
feat: Add error context to exceptions

Improved description on what is missing or what is the issue

wenenhoe committed:
refactor: Rename package to `storage`

wenenhoe committed:
feat: Add `find` command for keyword search

wenenhoe committed:
feat: Add support for storing dates

Replace `start` and `end` in `Deadline` and `Event` data format from
`String` to `LocalDate`

wenenhoe committed:
Merge pull request #1 from wenenhoe/branch-Level-8

feat: Add support for storing dates

wenenhoe committed:
refactor: Remove duplicated list printing code

Abstract into a function so it can be reused in printing matched tasks
and printing all tasks

wenenhoe committed:
feat: Add new data structure `Pair`

wenenhoe committed:
feat: Add parser to handle jumbled order args

wenenhoe committed:
refactor: Remove use of `parse` method

wenenhoe committed:
refactor: Utilise new parser to handle task args

wenenhoe committed:
Merge branch 'master' into branch-Refactor-Task-Parser

# Conflicts:
#	src/main/java/magus/task/variant/Deadline.java
#	src/main/java/magus/task/variant/Event.java

wenenhoe committed:
Merge pull request #2 from wenenhoe/branch-Refactor-Task-Parser

Refactor task parser

wenenhoe committed:
Merge remote-tracking branch 'origin/master' into branch-Level-9

wenenhoe committed:
feat: Add getters for `Deadline` and `Event`

wenenhoe committed:
feat: Add `UNKNOWN` task type and getEnum by name

For implementing task finding without task type specified

wenenhoe committed:
feat: Add new `TaskFinder` class

Abstract extended `find` functionality into separate class to handle
filtering by different task types and date searches

wenenhoe committed:
feat: Integrate `TaskFinder` into `TaskManager`

wenenhoe committed:
Merge pull request #3 from wenenhoe/branch-Level-9

Branch level 9

wenenhoe committed:
fix(bug): Fix issue with `getArgPositions`

* When `isIncludeFirstParam` is `false`, `endIndex` not obtained correctly
as the `for` loop just skips to the end
* Fix is applied to allow endIndex to be set in the next iteration of the
for loop

wenenhoe committed:
feat: Add `toString` method to `Pair`

Allow for better string output when debugging instances of this class

wenenhoe committed:
test: Add removal of saved task list file

wenenhoe committed:
test: Update test cases for new inputs and outputs

wenenhoe committed:
fix(bug): Handle `DateTimeParseException`

When adding tasks

wenenhoe committed:
test: Update tests to include erroneous inputs

wenenhoe committed:
feat: Add new exception for unknown arguments

wenenhoe committed:
feat: Add checks for detecting unknown argument

Console arguments are screened for unexpected arguments

wenenhoe committed:
feat: Handle new exception in consumer code

Code that utilises the console parser need to handle the new exception

wenenhoe committed:
test: Add test cases for unknown argument

wenenhoe committed:
feat: Handle non numerical input for commands

* `mark`
* `unmark`
* `delete`

wenenhoe committed:
test: Add test cases for non numerical inputs

* `mark`
* `unmark`
* `delete`

wenenhoe committed:
Merge pull request #4 from wenenhoe/branch-Test-All-Features

Branch test all features

wenenhoe committed:
Merge branch 'master' into branch-A-JavaDoc

wenenhoe committed:
docs: Add docs for `Pair`

wenenhoe committed:
docs: Add docs for `Task`

wenenhoe committed:
docs: Add docs for exception classes

wenenhoe committed:
style: Rename constants for convention

wenenhoe committed:
docs: Add docs for `FileIo`

wenenhoe committed:
refactor: Store split task info as member

wenenhoe committed:
docs: Add docs for task storage `Parser`

wenenhoe committed:
style: Remove unused import

wenenhoe committed:
refactor: Abstract to static method in Parser

For obtaining the argument value and checking if it is empty

wenenhoe committed:
style: Remove spacing from varargs typing

wenenhoe committed:
docs: Add docs for `Deadline`

wenenhoe committed:
docs: Add docs for `Event`

wenenhoe committed:
docs: Add docs for enums and `Magus`

wenenhoe committed:
docs: Add docs for `Todo`

wenenhoe committed:
docs: Add docs for `CommandManager`

wenenhoe committed:
docs: Add docs for `Console`

wenenhoe committed:
refactor: Remove unused method

wenenhoe committed:
docs: Add docs for `TaskManager`

wenenhoe committed:
docs: Add docs for console `Parser`

wenenhoe committed:
Merge pull request #5 from wenenhoe/branch-A-JavaDoc

Branch a java doc

wenenhoe committed:
docs: Update user guide

wenenhoe committed:
Merge pull request #7 from wenenhoe/branch-A-UserGuide

docs: Update user guide

okkhoy commented on src/main/java/magus/Magus.java:
```diff
@@ -0,0 +1,34 @@
+            }
+
+            // Exit when bye command issued
+            isExitProgram = parser.getCommand() == Command.BYE;
```
better to explicitly write a if block here

wenenhoe replied:
Resolved in 1b5c926

okkhoy commented on src/main/java/magus/console/CommandManager.java:
```diff
@@ -0,0 +1,79 @@
+ */
+public class CommandManager {
+    /**
+     * Takes the parsed user input and process the relevant command and additional inputs specified
```
better: `Processes the input ...`

wenenhoe replied:
Resolved in 84d4e90

okkhoy commented on src/main/java/magus/console/Parser.java:
```diff
@@ -0,0 +1,253 @@
+     * @param input User input from console
+     */
+    public Parser(String input) {
+        // Parse first word to get command
+        String firstWord = input;
```
misleading comment. you are assigning the entire input to `firstWord`.
additionally, `firstWord` is being used to manipulate the input. better to have a temporary variable to make it explicit.

wenenhoe replied:
Resolved in 8269dc2

okkhoy commented on src/main/java/magus/console/Parser.java:
```diff
@@ -0,0 +1,253 @@
+    private boolean hasUnknownArgument(String[] keywordArgs) {
+        String pattern = "^/.+$";
+        List<String> keywordArgsList = List.of(keywordArgs);
+        List<String> unknownArgsList = additionalInputList.stream()
+                .filter((input) -> input.matches(pattern)
+                        && !keywordArgsList.contains(input))
+                .collect(Collectors.toList());
```
😎

okkhoy commented on src/main/java/magus/task/TaskFinder.java:
```diff
@@ -0,0 +1,218 @@
+        case TODO:
+            try {
+                filteredTaskList = filterTodos(subparser);
+            } catch (ArgumentNotFoundException ignored) {
+                // break out and try filter by description
+                break;
```
is there a better way to do this rather than try--catch?
in general, it is not a good practice to have empty catch blocks

wenenhoe replied:
Resolved in 1c307c3

okkhoy reviewed (commented):
In general, well written code! 💯 
There are some minor things that I have pointed out. See if you can update the code before the submission.

wenenhoe committed:
docs: Reword javadoc for improved clarity

wenenhoe committed:
refactor: Improve clarity of `firstWord` usage

Use it as a temporary variable to obtain first word

wenenhoe committed:
refactor: Improve code readability

With explicit if condition

wenenhoe committed:
docs: Add docs for `TaskFinder`

wenenhoe committed:
Merge pull request #8 from wenenhoe/branch-Refactor-Add-Final-Touches

Branch refactor add final touches

wenenhoe committed:
test: Update outdated expected test result

wenenhoe committed:
Merge pull request #9 from wenenhoe/branch-Update-Test-Result

test: Update outdated expected test result

wenenhoe committed:
fix: Use `null` return if error

Ensure only error message printed

wenenhoe committed:
refactor: Handle exceptions instead of throw

Return null if any exception encountered

wenenhoe committed:
refactor: Replace duplicated functionality

`filterTodo` functions the same as `filterTasksByDescription` for `TODO`

wenenhoe committed:
chore: Remove unused import

wenenhoe committed:
refactor: Add filter by description

`filterEvents` filters by description as last option

wenenhoe committed:
refactor: Allow filter by description

`filterDeadlines` can filter by description

wenenhoe committed:
test: Update expected test result

wenenhoe committed:
Merge pull request #10 from wenenhoe/branch-Fix-Console-Parser

Branch fix console parser
